data/lua/as_text.lua

```diff
@@ -0,0 +1,66 @@
+-- NOTE: the string output from here is intended solely as an aid in debugging; it should never be taken and used as an input to anything else.
+
+local delete_chars = string.char(00, 01, 02, 03, 04, 05, 06, 07, 08, 09, 10, 11, 12, 13, 14, 15,
+	16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31) -- https://www.ietf.org/rfc/rfc4627.txt
+local delete_regexp = "[" .. delete_chars .. "]"
+
+-- escaping takes 3/4 of the time, but we can't avoid it...
+local function escape(str)
+	str = string.gsub(str, delete_regexp, "")
+	str = string.gsub(str, '\\', '\\\\')
+	return string.gsub(str, '"', '\\"')
+end
+
+local function add_table_key(obj, buffer)
+	local _type = type(obj)
+	if _type == "string" then
+		buffer[#buffer + 1] = escape(obj)
+	elseif _type == "number" then
+		buffer[#buffer + 1] = obj
+	elseif _type == "boolean" then
+		buffer[#buffer + 1] = tostring(obj)
+	else
+		buffer[#buffer + 1] = '???' .. _type .. '???'
+	end
+end
+
+local function format_any_value(obj, buffer)
+	local _type = type(obj)
+	if _type == "table" then
+		buffer[#buffer + 1] = '{'
+		buffer[#buffer + 1] = '"' -- needs to be separate for empty tables {}
+		for key, value in pairs(obj) do
+			add_table_key(key, buffer)
+			buffer[#buffer + 1] = '":'
+			format_any_value(value, buffer)
+			buffer[#buffer + 1] = ',"'
+		end
+		buffer[#buffer] = '}' -- note the overwrite
+	elseif _type == "string" then
+		buffer[#buffer + 1] = '"' .. escape(obj) .. '"'
+	elseif _type == "boolean" or _type == "number" then
+		buffer[#buffer + 1] = tostring(obj)
+	elseif _type == "userdata" then
```

Comment:
Should there be a case for translatable strings in the list? I think they had their own type or they are userdata.

Reply:
According to [here](https://wiki.wesnoth.org/LuaAPI#Wesnoth_Types) they'd be userdata.